app/src/main/java/org/wikipedia/notifications/NotificationPresenter.kt

```diff
@@ -1,37 +1,57 @@
 package org.wikipedia.notifications
 
-import android.app.NotificationManager
+import android.Manifest
 import android.app.PendingIntent
 import android.content.Context
 import android.content.Intent
+import android.content.pm.PackageManager
 import android.graphics.*
 import android.net.Uri
+import android.os.Build
+import androidx.activity.result.ActivityResultLauncher
 import androidx.annotation.ColorRes
 import androidx.annotation.DrawableRes
 import androidx.appcompat.view.ContextThemeWrapper
 import androidx.core.app.NotificationCompat
+import androidx.core.app.NotificationManagerCompat
 import androidx.core.app.PendingIntentCompat
 import androidx.core.app.RemoteInput
 import androidx.core.content.ContextCompat
-import androidx.core.content.getSystemService
 import androidx.core.graphics.applyCanvas
 import androidx.core.graphics.createBitmap
 import org.wikipedia.Constants
 import org.wikipedia.R
 import org.wikipedia.WikipediaApp
+import org.wikipedia.auth.AccountUtil
 import org.wikipedia.dataclient.WikiSite
 import org.wikipedia.diff.ArticleEditDetailsActivity
 import org.wikipedia.notifications.db.Notification
 import org.wikipedia.page.PageTitle
 import org.wikipedia.richtext.RichTextUtil
+import org.wikipedia.settings.Prefs
 import org.wikipedia.talk.TalkTopicsActivity
 import org.wikipedia.theme.Theme
 import org.wikipedia.util.*
 import org.wikipedia.util.log.L
-import java.util.*
+import java.util.Locale
+import java.util.concurrent.TimeUnit
 
 object NotificationPresenter {
 
+    private var lastPermissionRequestTime = 0L
+
+    fun maybeRequestPermission(context: Context, launcher: ActivityResultLauncher<String>) {
+        val millisSinceLastRequest = System.currentTimeMillis() - lastPermissionRequestTime
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
+            !Prefs.isInitialOnboardingEnabled &&
+            AccountUtil.isLoggedIn &&
+            (millisSinceLastRequest > TimeUnit.HOURS.toMillis(1) || millisSinceLastRequest < 0) &&
+            ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
+            launcher.launch(Manifest.permission.POST_NOTIFICATIONS)
+            lastPermissionRequestTime = System.currentTimeMillis()
```

Comment:
If I click on "Don't allow" and close the app, and next time the permission check dialog will show up again, if I click "Don't allow" again, it will not show up after the second deny. 

It looks like the system saves the preference somewhere in the device, just wondering, would it be possible to use that system preference in this logic so we do not have to show the dialog twice?

Reply:
> It looks like the system saves the preference somewhere in the device, just wondering, would it be possible to use that system preference in this logic so we do not have to show the dialog twice?

Why wouldn't we want to show the dialog twice? This would be the same as any other Permissions workflow, e.g. asking for location permissions in Places, asking for storage permissions when downloading a file. All of those can be denied, and will be asked again.

Reply:
Got it, it makes sense to me as long as it aligns with the other permission checks.